hugr-py/src/hugr/hugr.py

```diff
@@ -0,0 +1,313 @@
+from dataclasses import dataclass, replace
+
+from collections.abc import Collection, Mapping
+from typing import Iterable, Sequence, Protocol, Generic, TypeVar, overload
+
+from hugr.serialization.serial_hugr import SerialHugr
+from hugr.serialization.ops import BaseOp, OpType as SerialOp
+import hugr.serialization.ops as sops
+from hugr.serialization.tys import Type
+from hugr.utils import BiMap
+
+
+@dataclass(frozen=True, eq=True, order=True)
+class Port:
+    node: "Node"
+    offset: int
+    sub_offset: int = 0
+
+
+@dataclass(frozen=True, eq=True, order=True)
+class InPort(Port):
+    pass
+
+
+class ToPort(Protocol):
+    def to_port(self) -> "OutPort": ...
```

Comment:
What is the definition of a `ToPort`? It's implemented for `OutPort` and `Node`, but not for `Port` nor `InPort`.

Reply:
should really be `ToOutPort`, used so you can use a node in place of the single out port of the node (common for single return ops)